### DISCUSSION TITLE
Kernel configuration oddities

### DISCUSSION THREAD

**Discussion by MaxKellermann** [Openvario/meta-openvario#185] [General]
I tried to make sense of the OpenVario kernel configuration (https://github.com/Openvario/meta-openvario/blob/hardknott/recipes-kernel/linux/linux-mainline/arm/defconfig); it was written in a rather cumbersome way: instead of using `*.cfg` deltas to configure individual aspects in a modular way, like Yocto kernel recipes are supposed to do it, there's a full copy of the kernel configuration which pretents to be the "defconfig". Of course, that's very difficult to port to new kernels, and makes it hard to support different hardware platforms. (I stumbled upon that because I have no Cubieboard 2, and wanted to quickly port it to my Cubieboard 1 so I can test OpenVario builds there.)

So I converted the kernel configuration into a delta to the vanilla defconfig, and fond some oddities:
- why is the kernel modular? You could argue that some rarely used USB device drivers should be modules (though I disagree - and some are modular, and others are static), but lots of basic things like cpufreq governors, the GPU drivers, USB host drivers, and a few crypto modules are modular as well - and these are always loaded. Why modular?
- why module unloading?
- what use is the Contiguous Memory Allocator (`CONFIG_CMA`) for OpenVario?
- why CAN support? Does OpenVario have a CAN bus?
- no iptables support, even though connman cannot be built without iptables support (of course, OpenVario doesn't need a networking firewall - but connman insists... I don't like connman anyway)
- Dallas 1-wire - does OpenVario have that?
- DVB and V4L - what, can you watch TV on your OpenVario?
- USB printer support? Who has a printer in their glider??
- USB modem? If there's no 5G when you're flying, you use a USB modem, connected to ... which landline phone?
- Industrial I/O? Does OpenVario have that?
- Magic Sysrq?
- debugfs?
- stacktrace?
- ftrace?
- why ACL and xattr support?

These options are disabled in the default kernel config - they were explicitly enabled for the OpenVario custom kernel config. That's why I'm asking - who thought these options were useful for OpenVario?

They are part of the "initial commit" (6d9a70403aee896587df36c4d2f52cf70988401c) made by a guy called "build" (on host "build.fritz.box", huh??) just 3 years ago in 2019. I thought OpenVario was much older than 2019 - why is all the commit history of meta-openvario before 2019 lost?

(Related PR: https://github.com/Openvario/meta-openvario/pull/186)

**Comment by linuxianer99** [comment]
Hey @MaxKellermann 
I try to clarify some of your points ..

- Initial Commit
Think this was the switch to the mainline kernel ... but i am not sure

- Delta config vs. full defconfig
That is for historical reason. We started a long time ago ... so as i mentioned in another post before, there was not that deep knowledge about all the software stuff .. So i worked approx 3 years to get the first image (including XCSoar and the GLES stuff) running on the OV hardware. During these time my experience improved a lot, so i would implement things in a complete other way the i did in the past. All these "horrible" things are just a leftover of the starting times of Openvario.

- Kernel config itself
There was as PR a few days ago, doing a new config and also checking the needed options, this was just started, but not finished yet. Your input here is very welcome and i agree with you, that most of the users will never connect a printer in the glider ;-) There is big room for improvement ...

- Commit history
We started using a redmine server and Subversion in the very beginning. It was very cumbersome to add other people for example ... So we decided to port all the stuff to github. We don't migrate the commit history of the SVN as far a i can remember.

Hope this explains a bit the background, where we come from and why lot of the things here are looking that "horrible" ;-)

**Reply by MaxKellermann** [reply]
>  We started using a redmine server and Subversion in the very beginning. It was very cumbersome to add other people for example ... So we decided to port all the stuff to github. We don't migrate the commit history of the SVN as far a i can remember.

The "master" branch goes back to 2014, but "warrior" only to 2019.
Losing commit history is very unfortunate. Commit history always contains a wealth of wisdom, even after 10 years or more.

**Reply by MaxKellermann** [reply]
For example: why is `sunxi.inc` commented out in all machine configurations, but those files have copies of `sunxi.inc`? What was the author thinking? I have no idea, I cannot imagine, and I cannot look it up in the commit history, because it was already that way in the initial commit.
So instead of digging through the commit history, I have to find the people who did this and ask them. In the end, it'll turn out that nobody has been thinking anything, and everything was just a quick hack, as always.

**Reply by linuxianer99** [reply]
What is the reason for offending people here? Everybody does his best and we do this all in our spare time ....

**Reply by MaxKellermann** [reply]
I don't mean to offend anyone - what exactly offends you?

**Reply by linuxianer99** [reply]
"nobody has been thinking anything" ... keiner denkt hier irgendwas ...

**Reply by MaxKellermann** [reply]
This just means "was found by try'n'error without knowing what they were doing". That's the kind of reply I got every time I asked "why did you do it this or that way".
There are more elegant ways to do stuff with Yocto, but if you commit a full kernel config to the repository, and if you copy big chunks of BitBake rules from meta-sunxi, because you don't know better, this seems to work at first, but will make your life very miserable later on, when you need to update the Yocto version. Like it happened recently.
On my dayjob Yocto repository, updating to Hardknott was changing just two lines.
There's nothing wrong with making your own life harder than necessary; I'm not here to help, not to complain.
The reason I ask questions "why is it done this or that way" is because knowing is important. When I see a non-default USB printer driver is explicitly enabled in the kernel configuration, that doesn't appear to make any sense - I could just post a patch to remove it. But maybe somebody enabled it for a good reason - maybe somebody really has a portable printer in his glider, and he installed CUPS from a repository ... And which problem was solved by not using `sunxi.inc`? I need to know that, or else a change I propose will break something.
So far, such questions have never led to a positive result. That's why I said "nobody has been thinking anything", because stuff just happened by accident, with nobody aware why they happened. That's, of course, a valid response - and it's a useful response for me. Because it means I can go on and remove kludges that only make your life uneasy.

**Reply by linuxianer99** [reply]
OK ... understand your point.

So let's improve the quality of the source as well as the way we use Yocto for getting a good and stable image :-)

**Reply by MaxKellermann** [reply]
An example I have just been trying to wrap my head around:
https://github.com/Openvario/meta-openvario/blob/074e7e4748d67fc19965dbc2beb92190bee8dfd2/conf/machine/openvario-7-CH070.conf#L37
What the ... this is an obviously malformed address! How can this ever work? What use is this?
The first time I could find this in git was from 2015 on branch "next" in commit https://github.com/Openvario/meta-openvario/commit/05ad8afe8563ae#diff-dcd5bbd27701ee5f34b98c5c987f05acb796e00b422798c17665afa74fd2aff3R29
.... but this commit is so crowded, and this particular change still doesn't appear to make sense. I guess that this bad value was created in 2015 and from there, it was copied everywhere by everybody, and nobody ever understood why it is that way.
I don't say this to criticize anybody - I want to understand why it was added - always thinking that maybe I am wrong and missing something! If I know it really makes no sense, it is simple to remove it.

**Reply by linuxianer99** [reply]
It seems it is never used anymore ;-)
Think this is a leftover from sunxi-based kernel. This was used before sunxi was available in the mainline.
I agree with you, that it has to be cleaned up in the past .. but as is mentioned before, i had also no big experience in 2015 with Yocto, so everybody was happy, as a picture appears on the screen ... I know, I repeat me ;-)

But I think as rule of thumb, most of the strange things in the code are mistakes or dirty hacks. 
So if you find any more, don't waste your time thinking about the sense ;-) maybe just make a PR how it can be made better :-)

**Comment by kedder** [comment]
@MaxKellermann I think most of configuration you have listed are there not by conscious decision, but because it was like that by default or by accident and nobody complained. 

It makes sense for USB device drivers to be in modules because people connect various devices and there is no particular list of what can be connected to openvario. If someone wants to connect a printer, why would we prevent that? Note that not all devices are useful while you're flying, some are useful when you're on the ground and sending your logs to the scorer (for example).

For the other devices and features you listed - in my opinion we don't have to keep them enabled. I don't know anything that would depend on them.

**Reply by MaxKellermann** [reply]
> @MaxKellermann I think most of configuration you have listed are there not by conscious decision, but because it was like that by default or by accident and nobody complained.

There are lots of default options which are useless for OpenVario (and I'll submit a PR to remove them), but the options I posted here are not default options - they were explicitly enabled. I was asking because I wanted to know which ones of those can be disabled - maybe somebody really wanted to use a USB printer ...

> It makes sense for USB device drivers to be in modules

None of what you wrote explains why those are modules.

Modular kernels are hard to update; the kernel is on the boot partition, but the modules are on the root partition, and both must match exactly (not just the kernel version, but also many build options). This is fragile, so there must be a good reason to build a modular kernel. I see none, but I thought maybe somebody knows a good reason I don't see.

> If someone wants to connect a printer, why would we prevent that?

Does that mean you would like to keep the USB printer driver? Anybody who wishes to have certain very exotic modules can build their own kernel; it's open source after all. But I'm not convinced that a vario should have a USB printer driver, just in case somebody wants to print something. (OpenVario doesn't even have userspace printing support - but the same way you could argue that the next firmware build should ship with CUPS...)

**Reply by kedder** [reply]
> but the options I posted here are not default options - they were explicitly enabled

Again, the only reason I can think of is by accident. I don't know any specific reason why most of these options are enabled.

That said, please keep in mind that Openvario is a DIY electroincs project. People do creative things with their devices.We should not limit their creativity without a really good reason.

> None of what you wrote explains why those are modules.

Having them as modules means we don't have to load all the possible usb drivers into the memory. It also means those drivers will not conflict with each other. This is also a set up that is known to work. All desktop linux kernels (I know) are modular, and they are getting upgraded without problems.  Modular kernels are in production since forever and keeping kernel and modules in sync is a long time solved problem.

Also, linux kernel is not something we have to update very often. Unless we are fixing the kernel itself (like we had to do couple of years ago), a kernel version can stay with whatever yocto version provides.

> Does that mean you would like to keep the USB printer driver?

I don't have any strong opinion about the printer driver specifically. However, is there any good reason to explicitly disable it? Should we review each an every driver and decide whether it makes sense to have it on OV or not? I don't think so. It doesn't cost us anything to provide all the USB drivers that linux supports. However, shipping CUPS with OV and making sure printing actually works will cost us a lot of effort, therefore I don't think we should do that, since noone has found a good use for it so far.

**Comment by mihu-ov** [comment]
>Dallas 1-wire - does OpenVario have that?

Yes, we have a 1-wire interface on Openvario. It uses a DS2482-100 I2C to 1-Wire bridge, don´t know if this requires the 1-Wire kernel module or not. Maybe @linuxianer99 knows more about this bridge?

**Reply by linuxianer99** [reply]
Yes, i think this is used by Openvario (on sensorboard)

**Reply by MaxKellermann** [reply]
OK, when we re-add 1-wire support to the kernel config, the `.cfg` snippet should have a comment explaining this piece.

**Reply by MaxKellermann** [reply]
https://github.com/Openvario/meta-openvario/pull/222
